Content.Server/Automod/AutomodManager.cs

```diff
@@ -0,0 +1,260 @@
+            if (regexMatches.Count == 0)
+                continue;
+
+            var textMatches = new Dictionary<string, int>();
+            foreach (Match match in regexMatches)
+            {
+                var str = match.ToString();
+                textMatches.Add(str, match.Index);
+            }
+
+            if (!_protoMan.TryIndex<AutomodActionGroupPrototype>(filter.ActionGroup, out var censorGroup))
+            {
+                _log.Error($"AutomodActionGroupPrototype \"{filter.ActionGroup}\" not found.");
+                continue;
+            }
+
+            var skip = false;
+            foreach (var censorAction in censorGroup.AutomodActions)
+            {
+                if (!censorAction.Skip(inputText, textMatches))
+                    continue;
+
+                skip = true;
+                break;
+            }
+            if (skip)
+                continue;
+
+            var displayName = GetDisplayName(filter);
+
+            foreach (var censorAction in censorGroup.AutomodActions)
+            {
+                blocked &= censorAction.RunAction(session, inputText, textMatches, filter, displayName, _entMan);
+            }
+        }
+
+        return blocked;
+    }
+
+    private string GetDisplayName(AutomodFilterDef filter)
+    {
+        return filter.Id != null ? $"{filter.DisplayName} (automod#{filter.Id})" : $"{filter.DisplayName} (automod#na)";
+    }
+
+    #endregion
+
+    #region Reload
+
+    public async void ReloadAutomodFilters()
+    {
+        _regexFilters.Clear();
+
+        var filters = await _db.GetAllAutomodFiltersAsync();
+
+        foreach (var filter in filters)
+        {
+            AddFilter(filter);
+        }
+    }
+
+    #endregion
+
+    #region Create
+
+    public async void CreateFilter(string pattern,
+        AutomodFilterType filterType,
+        string actionGroup,
+        AutomodTarget targets,
+        string name)
+    {
+        CreateFilter(new AutomodFilterDef(pattern, filterType, actionGroup, targets, name));
+    }
+
+    public async void CreateFilter(AutomodFilterDef automod)
+    {
+        automod = await _db.AddAutomodFilterAsync(automod);
+        AddFilter(automod);
+    }
+
+    private void AddFilter(AutomodFilterDef automod)
+    {
+        foreach (AutomodTarget targetFlag in Enum.GetValues(typeof(AutomodTarget)))
+        {
+            if (targetFlag == AutomodTarget.None)
+                continue;
+            if (!automod.TargetFlags.HasFlag(targetFlag))
+                continue;
+
+            if (automod.FilterType == AutomodFilterType.Regex)
+            {
+                if (!_regexFilters.TryGetValue(targetFlag, out var list))
+                {
+                    list = new Dictionary<Regex, AutomodFilterDef>();
+                    _regexFilters.Add(targetFlag, list);
+                }
+
+                list.Add(new Regex(automod.Pattern), automod);
+            }
+            // TODO other filter types
```

Comment:
If this is intended to be merged before this TODO is finished, this needs to log osmething.